src/sentry/db/models/queryset.py

```diff
@@ -0,0 +1,102 @@
+from __future__ import absolute_import
+
+__all__ = ('BoundDateQuerySet', 'BoundDateTimeQuerySet',
+           'BoundQuerySet', 'BoundValuesQuerySet')
+
+import six
+
+from django.db.models.query import DateQuerySet, DateTimeQuerySet, QuerySet, ValuesQuerySet
+
+
+def bound_queryset(cls, prefix='Bound'):
+    class new_cls(cls):
+        def __init__(self, *args, **kwargs):
+            # this must be optional for _clone
+            self._binding_criteria_fn = kwargs.pop('binding_criteria_fn', None)
+            self._constraints_applied = False
+            cls.__init__(self, *args, **kwargs)
+
+        def __getitem__(self, k):
+            if self._constraints_applied:
+                qs = self
+            else:
+                qs = self._clone_with_constraints()
+            return cls.__getitem__(qs, k)
+
+        # XXX(dcramer): this cloning stuff is expensive -- we could,
+        # alternatively, just bind an attribute
+        def _clone(self, klass=None, unconstrained_unsafe=False, *args, **kwargs):
+            if klass and not unconstrained_unsafe:
+                try:
+                    klass = CLASS_MAPPING[klass]
+                except KeyError:
+                    raise NotImplementedError(
+                        'Unable to clone to a bound type for {:r}'.format(klass))
+            rv = cls._clone(self, klass=klass, *args, **kwargs)
+            rv._constraints_applied = self._constraints_applied
+            rv._binding_criteria_fn = self._binding_criteria_fn
+            return rv
+
+        # XXX(dcramer): if we could find a way to clear these parameters (which is tricky)
```

Comment:
i kept this comment in case anyone ever looks at this code and asks "why cant we use _setup_query"